src/jit/function-builder.cc

```diff
@@ -635,6 +635,83 @@ void FunctionBuilder::EmitUnsignedTruncation(TR::IlBuilder* b, const uint8_t* pc
   stack->Push(new_value);
 }
 
+template <typename ToType, typename FromType>
+void FunctionBuilder::EmitSaturatingTruncation(TR::IlBuilder* b, VirtualStack* stack) {
+  static_assert(std::is_floating_point<FromType>::value, "FromType in EmitTruncation call must be a floating point type");
+
+  auto* value = stack->Pop();
+  auto* result = b->Const(static_cast<ToType>(0));
+
+  TR::IlBuilder* non_nan_path = nullptr;
+
+  b->IfThen(&non_nan_path, b->EqualTo(EmitIsNan<FromType>(b, value), b->ConstInt32(0)));
+
+  TR::IlBuilder* too_high_path = nullptr;
+  TR::IlBuilder* not_too_high_path = nullptr;
+
+  non_nan_path->IfThenElse(&too_high_path, &not_too_high_path,
+  non_nan_path->           GreaterThan(value,
+  non_nan_path->                       Const(static_cast<FromType>(std::numeric_limits<ToType>::max()))));
+  too_high_path->StoreOver(result, too_high_path->Const(std::numeric_limits<ToType>::max()));
+
+  TR::IlBuilder* too_low_path = nullptr;
+  TR::IlBuilder* not_too_low_path = nullptr;
+
+  not_too_high_path->IfThenElse(&too_low_path, &not_too_low_path,
+  not_too_high_path->           LessThan(value,
+  not_too_high_path->                    Const(static_cast<FromType>(std::numeric_limits<ToType>::lowest()))));
+  too_low_path->StoreOver(result, too_low_path->Const(std::numeric_limits<ToType>::lowest()));
+
+  auto* target_type = toIlType<ToType>(b->typeDictionary());
+
+  // this could be optimized using templates or constant expressions,
+  // but the compiler should be able to simplify this anyways
+  auto* new_value = std::is_unsigned<ToType>::value ? not_too_low_path->UnsignedConvertTo(target_type, value)
```

Comment:
This part is copied verbatim from `EmitTruncation`... I think the idea was that once the unsigned truncation opcodes were properly implemented upstream, the special versions of these functions for unsigned truncation to 32-bit integers would go away since the special behaviour they implement would no longer be needed.